tests/bluechi_test/bluechilib/bluechictl.py

```diff
@@ -0,0 +1,129 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+import copy
+import signal
+import subprocess
+import threading
+import time
+import tempfile
+from typing import List, Pattern, Set, Tuple
+
+
+class FileFollower:
+    def __init__(self, file_name):
+        self.pos = 0
+        self.file_name = file_name
+        self.file_desc = None
+
+    def __enter__(self):
+        self.file_desc = open(self.file_name, mode='r')
+        return self
+
+    def __exit__(self, exception_type, exception_value, exception_traceback):
+        if (exception_type):
+            print(f"Exception raised: excpetion_type='{exception_type}', "
+                  f"exception_value='{exception_value}', exception_traceback: {exception_traceback}")
+        if self.file_desc:
+            self.file_desc.close()
+
+    def __iter__(self):
+        while self.new_lines():
+            self.seek()
+            line = self.file_desc.read().split('\n')[0]
+            yield line
+
+            self.pos += len(line) + 1
+
+    def seek(self):
+        self.file_desc.seek(self.pos)
+
+    def new_lines(self):
+        self.seek()
+        return '\n' in self.file_desc.read()
+
+
+class BluechiCtlBackgroundRunner:
+    def __init__(self, command: List, expected_patterns: Set[Pattern]):
+        self.command = command
+        self.expected_patterns = copy.copy(expected_patterns)
+        self.thread = threading.Thread(target=self.process_events)
+        self.failsafe_thread = threading.Thread(target=self.timeout_guard, daemon=True)
+        self.all_found = False
+        self.finished = False
+        self.bluechictl_proc = None
+
+    def process_events(self):
+        print("Starting process_events")
+        with tempfile.NamedTemporaryFile() as out_file:
+            try:
+                self.bluechictl_proc = subprocess.Popen(self.command, stdout=out_file, bufsize=1)
+
+                with FileFollower(out_file.name) as bluechictl_out:
+                    while self.bluechictl_proc.poll() is None and not self.finished:
+                        for line in bluechictl_out:
+                            print(f"Evaluating line '{line}'")
+                            match = None
+                            for pattern in copy.copy(self.expected_patterns):
```

Comment:
Just evaluating list of matches is not enough for more complicated use cases such as

https://github.com/eclipse-bluechi/bluechi/blob/main/tests/tests/tier0/bluechi-monitor-service/python/monitor.py
https://github.com/eclipse-bluechi/bluechi/blob/main/tests/tests/tier0/bluechictl-monitor-unit-created-removed/python/monitor.py

I suggest to use following class to evaluate incoming lines:

```python
class EventEvaluator:

def process_line(self, line: str) -> None:

def processing_finished(self) -> Bool:
```

Then each test could implement its own ancestor of this class.

Reply:
I thought about something like this, just thought I would first implement what is needed for `metrics listen`
I will add it.
Though I think in at least the `bluechi-monitor-service` test can be done with the existing processor as well. You can listen for specific outputs during specific commands quite easily, and it would remove the need of complicated event evaluator

Reply:
Maybe, but having a separate class with event evaluation makes it more understandable, versatile and also more separated from the BlueChiCtlBackgroundRunner class implementation ...

Reply:
You are right, the code actually looks much nicer with it.